flambe/nn/transformer.py

```diff
@@ -0,0 +1,532 @@
+        for p in self.parameters():
+            if p.dim() > 1:
+                xavier_uniform_(p)
+
+
+class TransformerEncoderLayer(Module):
+    """TransformerEncoderLayer is made up of self-attn and feedforward.
+
+    This standard encoder layer is based on the paper "Attention Is
+    All You Need". Ashish Vaswani, Noam Shazeer, Niki Parmar,
+    Jakob Uszkoreit, Llion Jones, Aidan N Gomez, Lukasz Kaiser, and
+    Illia Polosukhin. 2017. Attention is all you need. In Advances in
+    Neural Information Processing Systems, pages 6000-6010. Users may
+    modify or implement in a different way during application.
+
+    """
+
+    def __init__(self,
+                 d_model: int,
+                 nhead: int,
+                 dim_feedforward: int = 2048,
+                 dropout: float = 0.1,
+                 use_sru: bool = False) -> None:
+        """Initialize a TransformerEncoderLayer.
+
+        Parameters
+        ----------
+        d_model : int
+            The number of expected features in the input.
+        n_head : int
+            The number of heads in the multiheadattention models.
+        dim_feedforward : int, optional
+            The dimension of the feedforward network (default=2048).
+        dropout : float, optional
+            The dropout value (default=0.1).
+        use_sru: bool, optional
+            If true, replaces the FFN with an SRU.
+
+        """
+        super(TransformerEncoderLayer, self).__init__()
+
+        self.self_attn = MultiheadAttention(d_model, nhead, dropout=dropout)
+
+        if use_sru:
+            self.sru = SRU(d_model, dim_feedforward, num_layers=1, dropout=dropout)
+        else:
+            self.dropout = nn.Dropout(dropout)
+            self.linear1 = nn.Linear(d_model, dim_feedforward)
+
+        self.linear2 = nn.Linear(dim_feedforward, d_model)
+
+        self.norm1 = nn.LayerNorm(d_model)
+        self.norm2 = nn.LayerNorm(d_model)
+        self.dropout1 = nn.Dropout(dropout)
+        self.dropout2 = nn.Dropout(dropout)
+
+    def forward(self,  # type: ignore
+                src: torch.Tensor,
+                src_mask: Optional[torch.Tensor] = None,
+                src_key_padding_mask: Optional[torch.Tensor] = None) -> torch.Tensor:
+        """Pass the input through the endocder layer.
+
+        Parameters
+        ----------
+        src: torch.Tensor
+            The seqeunce to the encoder layer (required).
+        src_mask: torch.Tensor, optional
+            The mask for the src sequence (optional).
+        src_key_padding_mask: torch.Tensor, optional
+            The mask for the src keys per batch (optional).
+
+        """
+        src2 = self.self_attn(src, src, src, attn_mask=src_mask,
+                              key_padding_mask=src_key_padding_mask)[0]
+        src = src + self.dropout1(src2)
+        src = self.norm1(src)
+        if self.use_sru:
+            src2 = self.linear2(self.sru(src))
```

Comment:
specify padding mask for SRU:
`self.sru(src, mask_pad=src_key_padding_mask)`

Reply:
great catch

Reply:
i believe it should be `src_key_padding_mask` instead of `1 - src_key_padding_mask` based on the doc string. and it should be Byte tensor already.